BEIMA.Backend/MongoService/BaseDevice.cs

```diff
@@ -0,0 +1,174 @@
+        }
+
+
+        [BsonId]
+        public ObjectId Id { get; set; }
+
+        [BsonElement("deviceTypeId")]
+        public ObjectId DeviceTypeId { get; set; }
+
+        [BsonElement("deviceTag")]
+        public string DeviceTag { get; set; }
+
+        [BsonElement("manufacturer")]
+        public string Manufacturer { get; set; }
+
+        [BsonElement("modelNum")]
+        public string ModelNum { get; set; }
+
+        [BsonElement("serialNum")]
+        public string SerialNum { get; set; }
+
+        [BsonElement("yearManufactured")]
+        public int YearManufactured { get; set; }
+
+        [BsonElement("notes")]
+        public string Notes { get; set; }
+
+        [BsonElement("fields")]
+        public BsonDocument Fields { get; set; }
+
+        [BsonElement("location")]
+        public BsonDocument Location { get; set; }
+
+        [BsonElement("lastModified")]
+        public BsonDocument LastModified { get; set; }
+
+        /// <summary>
+        /// Gets a BsonDocument that represents the current state of the BaseDevice object.
+        /// </summary>
+        /// <returns>BsonDocument that represents the current state of the BaseDevice object. Conforms to the schema located in BEIMA.DB.Schemas.</returns>
+        /// <exception cref="ArgumentNullException">Throws exception when any of the required fields are null.</exception>
+        public BsonDocument GetBsonDocument()
+        {
+            if (DeviceTag == null)
+            {
+                throw new ArgumentNullException($"BaseDevice - {nameof(DeviceTag)} is null");
+            }
+
+            if (Manufacturer == null)
+            {
+                throw new ArgumentNullException($"BaseDevice - {nameof(Manufacturer)} is null");
+            }
+
+            if (ModelNum == null)
+            {
+                throw new ArgumentNullException($"BaseDevice - {nameof(ModelNum)} is null");
+            }
+
+            if (SerialNum == null)
+            {
+                throw new ArgumentNullException($"BaseDevice - {nameof(SerialNum)} is null");
+            }
+
+            if (Notes == null)
+            {
+                throw new ArgumentNullException($"BaseDevice - {nameof(Notes)} is null");
+            }
+            if (Location.ElementCount == 0)
+            {
+                throw new ArgumentNullException("BaseDevice - SetLocation has not been called yet!");
+            }
+            if (LastModified.ElementCount == 0)
+            {
+                throw new ArgumentNullException("BaseDevice - SetLastModified has not been called yet!");
+            }
+            return this.ToBsonDocument();
+        }
+
+        /// <summary>
+        /// Adds a custom field to the "fields" object.
+        /// </summary>
+        /// <param name="key">The ObjectId of the linked field in the linked DeviceType.</param>
+        /// <param name="value">The actual value of the field.</param>
+        public void AddField(string key, dynamic value)
```

Comment:
This method adds a value for a custom field that was defined in the device type. Do we need a method to do that, but for a custom field on a single device? I think that was a scenario that Brian had described. Maybe this will be tackled in a future task, but thought I'd bring it up to make sure

Reply:
I think that for now, this is fine, but in the future, I think we would likely want to use some sort of interaction with a deviceType object when setting these fields. Or even just query the database to see of the field value is valid for the given device type.